pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -584,3 +593,191 @@ func (a *Actuator) nodeAddresses(host *bmh.BareMetalHost) ([]corev1.NodeAddress,
+	if machine.Status.NodeRef == nil {
+		return nil, errors.NewNotFound(corev1.Resource("ObjectReference"), machine.Name)
+	}
+
+	node := &corev1.Node{}
+	key := client.ObjectKey{
+		Name:      machine.Status.NodeRef.Name,
+		Namespace: machine.Status.NodeRef.Namespace,
+	}
+
+	if err := a.client.Get(ctx, key, node); err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+/*
+remediateIfNeeded will try to remediate unhealthy machines (annotated by MHC) by power-cycle the host
+The full remediation flow is:
+1) Power off the host
+2) Add poweredOffForRemediation annotation to the unhealthy Machine
+3) Delete the node
+4) Power on the host
+5) Wait for the node the come up
+6) Remove poweredOffForRemediation annotation and the annotation added by MAO to signal the machine is unhealthy
+ */
+func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1.Machine, baremetalhost *bmh.BareMetalHost) error {
+	if len(machine.Annotations) == 0 {
+		return nil
+	}
+
+	if _, needsRemediation := machine.Annotations[externalRemediationAnnotation]; !needsRemediation {
+		return nil
+	}
+
+	if _, poweredOffForRemediation := machine.Annotations[poweredOffForRemediation]; !poweredOffForRemediation {
+		if !hasPowerOffRequestAnnotation(baremetalhost) {
+			log.Printf("Found an unhealthy machine, requesting power off. Machine name: %s", machine.Name)
+			return a.requestPowerOff(ctx, baremetalhost)
+		}
+
+		//hold remediation until the power off request is fulfilled
+		if baremetalhost.Status.PoweredOn {
+			return nil
+		}
+
+		//we need this annotation to differentiate between unhealthy machine that
+		//needs remediation, and an unhealthy machine that just got remediated
+		return a.addPoweredOffForRemediationAnnotation(ctx, machine)
+	}
+
+	node, err := a.getNodeByMachine(ctx, machine)
+
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Printf("Failed to get Node from Machine %s: %s", machine.Name, err.Error())
+			return err
+		}
+	}
+
+	if node != nil && !baremetalhost.Status.PoweredOn {
+		log.Printf("Deleting Node %s associated with Machine %s", node.Name, machine.Name)
+		/*
+		Delete the node only after the host is powered off. Otherwise, if we would delete the node
+		when the host is powered on, the scheduler would assign the workload to other nodes, with the
+		possibility that two instances of the same application are running in parallel. This might result
+		in corruption or other issues for applications with singleton requirement. After the host is powered
+		off we know for sure that it is safe to re-assign that workload to other nodes.
+		 */
+		return a.deleteNode(ctx, node)
+	}
+
+	// node is deleted, we can power on the host
+	if !baremetalhost.Status.PoweredOn {
+		log.Printf("Requesting Host %s power on for Machine %s",
+			baremetalhost.Name, machine.Name)
+		return a.requestPowerOn(ctx, baremetalhost)
+	}
+
+	// node is now available again
+	if node != nil {
```

Comment:
I think I would have written this conditional in the opposite way, and returned early here instead of letting that case fall through to the end of the method. That would make the order of the states a bit more obvious. First we wait for the node to come back, then we delete the annotation as the final step.

Reply:
Yeah, it makes more sense. Changed.